source/slang/ir-ssa.cpp

```diff
@@ -0,0 +1,848 @@
+    // one by inserting a new block. Note that we do not try
+    // to break the edges while doing the initial walk, because
+    // that would change the CFG while we are walking it.
+
+    for (auto edgeUse : criticalEdges)
+    {
+        auto pred = (IRBlock*) edgeUse->user->parent;
+        assert(pred->op == kIROp_Block);
+
+        auto succ = (IRBlock*)edgeUse->usedValue;
+        assert(succ->op == kIROp_Block);
+
+        IRBuilder builder;
+        builder.sharedBuilder = &context->sharedBuilder;
+        builder.curFunc = globalVal;
+        builder.curBlock = pred;
+
+        // Create a new block that will sit "along" the edge
+        IRBlock* edgeBlock = builder.createBlock();
+
+        // The predecessor block should now branch to
+        // the edge block.
+        edgeUse->set(edgeBlock);
+
+        // The edge block should branch (unconditionally)
+        // to the successor block.
+        builder.curBlock = edgeBlock;
+        builder.emitBranch(succ);
+
+        // Insert the new block into the block list
+        // for the function.
+        //
+        // In principle, the order of this list shouldn't
+        // affect the semantics of a program, but we
+        // might want to be careful about ordering anyway.
+        edgeBlock->insertAfter(pred);
+    }
+}
+
+// Construct SSA form for a global value with code
+void constructSSA(ConstructSSAContext* context)
+{
+    // First, detect and and break any critical edges in the CFG,
+    // because our representation of SSA form doesn't allow for them.
+    breakCriticalEdges(context);
+
+
+    // Figure out what variables we can promote to
+    // SSA temporaries.
+    identifyPromotableVars(context);
+
+    // If none of the variables are promote-able,
+    // then we can exit without making any changes
+    if (context->promotableVars.Count() == 0)
+        return;
+
+    // We are going to walk the blocks in order,
+    // and try to process each, by replacing loads
+    // and stores of promotable variables with simple values.
+
+    auto globalVal = context->globalVal;
+    for (auto bb = globalVal->firstBlock; bb; bb = bb->nextBlock)
+    {
+        auto blockInfo = new SSABlockInfo();
+        blockInfo->block = bb;
+
+        blockInfo->builder.sharedBuilder = &context->sharedBuilder;
+        blockInfo->builder.curBlock = bb;
+        blockInfo->builder.curFunc = globalVal;
+        blockInfo->builder.insertBeforeInst = bb->lastInst;
+
+        context->blockInfos.Add(bb, blockInfo);
+    }
+    for (auto bb = globalVal->firstBlock; bb; bb = bb->nextBlock)
+    {
+        auto blockInfo = * context->blockInfos.TryGetValue(bb);
+        processBlock(context, bb, blockInfo);
+    }
+
+    // We need to transfer the logical arguments to our phi nodes
+    // from the phi nodes back to the predecessor blocks that will
+    // pass them in.
+    for (auto bb = globalVal->firstBlock; bb; bb = bb->nextBlock)
```

Comment:
I don't quite understand what this code is doing here. Is it trying to add implicit block parameters as a result of some vars from the predecessor of a block is used by a successeor of the block (but not used by this block itself)? In that case, shouldn't `readVarRec` calls made by `processBlock` already added a phi node to this block?

Reply:
I see that now. This is to replace the old terminator instructions to include the additional arguments for the downstream phi nodes. Now I need to understand how these arguments are arranged for different sucessor blocks. For example, a conditional block has three succesor blocks, do the terminator instruction stores arguments for the union of vars from all sucessors, or the arguments for each succesor block in order?

Reply:
Most of the code for SSA construction is straight out of the Braun et al. paper that I used as a reference. The loop you are seeing there is a final step that is required because our IR currently doesn't have an explicit `phi` instruction that goes in the successor block, and instead has the block define `param`s for which the predecessor block passes arguments.

Basically, instead of something like:

```
A:
    %t0 = ...
    br C

B:
    %t1 = ...
    br C

C:
    %t2 = phi(%t0, %t1)
    ...
```

Our IR encodes something like:

```
A:
    %t0 = ...
    br C(%t0)

B:
    %t1 = ...
    br C(%t1)

C(param %t2):
    ...
```

During the construction process, it is easiest to think in terms of building up a phi (%t2) with a list of its operands (%t0, %t1 - one for each predecessor of the block), and that is what the `PhiInfo` type is tracking. Once we've computed all the information, though, the encoding is that each conceptual phi becomes a parameter added to block `C`, and each operand becomes an *argument* added to one of the predecessor blocks of `C`. That last loop is the step where we transfer things from the `PhiInfo` structure and actually start modifying things in the IR structures.

Some of the constraints on our IR (e.g., that instructions are allocated for a fixed number of operands) make it hard to do those modifications in place while generating the phis. One option would be to have an explicit `phi` instruction that we use during that pass, and then either translate away (in a loop like the one I wrote), or just drop the idea of parameters on blocks and keep the phis around.

I think its in the commit message, but the main reason I prefer "blocks with parameters" over traditional phi nodes is that it gets rid of the dependency on the implicit ordering of predecessors (that is, to read a phi node you need to know the order of the predecessors *or* encode the predecessors into the phi). It is also conceptually cleaner in terms of representing the way that phi nodes implicitly all execute in parallel at the start of a block (you can't just interpret them as acting sequentially). It also cleans up the conceptually ugly bit that in SSA all defs dominate their uses *except* for some special cases around phis (e.g., a phi at the top of a loop can use a definition in the loop body, which it dominates).

I'm not wedded to that design choice, though. I'm also not wedded to the current approach for computing SSA form. I just used the paper that seemed easier to implement, but realistically we are going to need dominator trees really soon anyway, so once we've written that code we might want to just re-use it for SSA construction as well.

Reply:
> Now I need to understand how these arguments are arranged for different sucessor blocks. For example, a conditional block has three succesor blocks, do the terminator instruction stores arguments for the union of vars from all sucessors, or the arguments for each succesor block in order?

This is something that is maybe too subtle in the IR encoding right now. The only terminator that is allowed to pass arguments is the `unconditionalBranch`, and so it only needs to deal with the arguments for one destination block.

An important thing to note is that a phi/parameter is only ever needed for a block with multiple predecessors (a "join point" in the CFG). A block with only a single predecessor should never need/have parameters (and our SSA construction algorithm currently guarantees this - `readVarRec` has a special case for a block with a single predecessor).

The case you are worried about is when a block with multiple successors needs to pass arguments. If that situation arose, it would mean that we have a CFG edge from a block with multiple **successors** to a block with multiple **predecessors**. Such an edge is called a "critical edge."

Some compiler passes can deal with critical edges just fine, but one particular place where they can cause problems is when taking SSA code and converting it back to code with assignments. In particular, when you have a critical edge the assignments implied by your phi nodes need to happen "along the edge." The traditional way to solve that problem is to "break" any critical edges that cause problems (by inserting a new block along the edge).

So, my solution in the IR today is to break *all* critical edges before constructing SSA form. Once I've done that there is an invariant that any node with multiple successors (a conditional branch of switch) *cannot* branch to a block with multiple predecessors (which might have parameters).

We could add subsequent passes that eliminate trivial blocks whenever it would not break our SSA invariants. I'm assuming we'd want some level of control-flow optimization like that anyway (in particular we need to drop unreachable blocks before we start issuing errors about functions failing to return on all code paths).

It would be pretty easy to extend the `conditionalBranch` and its variants to support passing arguments if we really wanted (`switch` would be annoyingly harder), but I'm not convinced it would be worth the complexity compared to the simpler approach of breaking critical edges (which again, we probably need to support anyway when translating out of SSA form).